### ISSUE TITLE
Deploy on same bench

### ISSUE THREAD

**Issue by rmehta** [frappe/press#545]
Allow a pull-only deploy for fast updates on same bench where there are no schema migrations

**Comment by ankush** [comment]
This is not worth having at cost of immutable images. 👎 


Aditya already mentioned that they just need to figure out the parts which slow down image builds for such cases. Once that time is around 1-5 minute it won't matter much.

**Comment by rmehta** [comment]
> This is not worth having at cost of immutable images. 👎

Why? Religion?

**Comment by ankush** [comment]
Many reasons:

- This will require special handling:
   - If py change then Restart workers so they see new code
   - if JS change then build and relink symlinks (not just `bench build`)
- Takes away the basic idea behind containers. Images don't represent the running version anymore. Just destroying and recreating containers also doesn't work anymore. You'll get rolled back to the previous version.
- What do we show on the dashboard about which commit it's running commit? 

TL;DR: Immutability of images good. Call it religion or whatever.

**Comment by rmehta** [comment]
#1 is trivial - used to be there in Central
#2 we don't need to destroy and recreate containers anyways - and they can be tagged in the model
#3 the last updated commit for sure

TLDR: Religious thinking is not so good. In this case, saves a lot of wasteful processing (?)

Not an urgent feature, just leaving it out there

**Comment by ankush** [comment]
>  is trivial

It's not.

More cases:

1. In shared hosting - What if there is more than one site? Everyone gets updated cause one site wants to pull update on their container? 
2. If we are doing this then why not do it for non-pull builds also? 🥴

**Comment by rmehta** [comment]
> In shared hosting - What if there is more than one site? Everyone gets updated cause one site wants to pull update on their container?

Isn't that a problem with shared containers in general.

> If we are doing this then why not do it for non-pull builds also? 🥴

In pull-only, you don't need migrations

**Comment by ankush** [comment]
> Isn't that a problem with shared containers in general.

It's not currently a problem because whoever wants update just moves to next container.

**Comment by adityahase** [comment]
We can make pull-only deploys faster by doing `git pull` on top of the existing image, during the build process. 

1. Won't change any current flows
1. Will be cache friendly (Since only the last layer that includes the diff needs to be pushed)
1. Should save a lot of time (No `setup requirements`, no `build`)

**Comment by rmehta** [comment]
Reopening this. We should be able run any deploys on same bench.

Canary deploys make sense for benches with large number of sites. For smaller sites / self managed servers - the need to "fast deploy" is paramount to the development process. 

There should be an option to deploy on same bench instead of building new containers. This feels like a no brainer.

cc: @18alantom @deepeshgarg007

**Comment by 18alantom** [comment]
Delta Builds https://github.com/frappe/press/pull/1286 having pull updates have been tried, there were issues with it and now it's been disabled.

This is different from what Rushabh's suggesting which is to not build an image on a deploy and to just update the running container. Which is what App Patches do ([ref](https://frappecloud.com/docs/benches/app-patches)).

Delta Builds, aimed to build a new image where the only diff was the pull update. The overhead w.r.t what Rushabh's suggesting ideally would've been just the time taken: to 
1. Upload build context to the server
2. Build the image
3. Upload it to the registry
4. Download it to the server
5. Shift bench to the new one

Due to how docker works with it's layer caching steps 2, 3, and 4 wouldn't take as much time as one would expect. The layer diff would just be a few KB (depending on the pull diff). And so for trivial changes ideally the time for deploy to be live would be under a minute.

---

This unfortunately is not how things worked out. Since we can't accurately predict whether a layer would be present, there was high incidence of false pull-update (~60%):

![telegram-cloud-photo-size-5-6127375068494413245-y](https://github.com/user-attachments/assets/2c9d89f7-91ca-402e-8434-083fcba35f79)

That is, instead of `[building app latest]` we ended up doing `[building app old + pull update latest]` this caused amount of wasted time to be more than the amount of time saved by doing a delta update.

After configuring to having 2 build servers the false pull-update rate shot up to above 80%.

Not only did it waste time, there were numerous hard to debug issues cause we can't say with certainty if an update can just be a pull-update, and due to Delta Updates functionality clashing with bench caching behavior.

---

Rushabh's idea of just updating the running container could work by persisting the change by using `docker container commit` ([ref](https://docs.docker.com/reference/cli/docker/container/commit)).

But the issue here is resultant image from the container commit cannot be used by others. So no shareable cache is created for that specific update.

Everyone who has to deploy that update is running the update separately. This saves a single developers time, but cumulatively wastes everyone's time.

![Screenshot 2024-08-16 at 13 04 21](https://github.com/user-attachments/assets/8be0ccb1-62d2-46d6-8f1a-a529691efa10)

**Comment by rmehta** [comment]
For private benches, it makes more sense to just deploy on the same bench / container and update it.

For public benches, it makes more sense to create fresh benches and migrate sites one-by-one.

Let's not push canary deploys for everyone and give developers a choice.

**Comment by 18alantom** [comment]
Addressed in: https://github.com/frappe/press/pull/2079
